File: day_1/python/part_two.py
import re


NUMBER_MAPPINGS = {
    'one': '1',
    'two': '2',
    'three': '3',
    'four': '4',
    'five': '5',
    'six': '6',
    'seven': '7',
    'eight': '8',
    'nine': '9'
}
NUMBERS_REGEX: str = f"(?=({'|'.join(NUMBER_MAPPINGS.keys())}|{'|'.join(NUMBER_MAPPINGS.values())}))"

input_file: str = 'input.txt'
output: int = 0
input_lines = []
output_lines = []


def convert_to_numbers(input_string: str) -> str:
    input = input_string.lower()
    output_list = []
    
    numbers = list(re.finditer(NUMBERS_REGEX, input))
    for number in numbers:
        number = number.group(1)
        if number.isnumeric():
            output_list.append(number)
        else:
            output_list.append(NUMBER_MAPPINGS[number])

    return ''.join(output_list)

def validate(answers_file: str):
    
    with open(answers_file, 'r') as file:
        answers = file.read().split('\n')
        for i in range(len(answers)):
           if output_lines[i] != answers[i]:
                print(f"Discrepancy found on line {i}. Original: {input_lines[i]}, Output: {output_lines[i]}, Answer: {answers[i]}")


with open(input_file, 'r') as input:

    lines: list[str] = input.read().split('\n')
    input_lines = lines
    for line_string in lines:

        line = convert_to_numbers(line_string)
        if(len(line) < 1): 
            output_lines.append('')
            continue
        
        number: str = line[0] + line[-1]
        output_lines.append(number)
        try: 
            output += int(number)
        except ValueError as error:
            print(f"'{number}' was not a number: \n{error}")

#validate('bulk_test_answers.txt')

print(output)

File: day_1/rust/src/part_two.rs
use std::{fs, collections::HashMap};
use onig::*;

const INPUT_FILE: &str = "input.txt";
fn main() {
    let number_map: HashMap<&str, &str> = HashMap::from([
        ("one", "1"),
        ("two", "2"),
        ("three", "3"),
        ("four", "4"),
        ("five", "5"),
        ("six", "6"),
        ("seven", "7"),
        ("eight", "8"),
        ("nine", "9")
    ]);

    let text: String = fs::read_to_string(INPUT_FILE)
        .unwrap_or_else(|_| panic!("file path '{}' should be a valid text file", INPUT_FILE));

    //println!("Text:\n{}\n", text);
    
    let result: u32 = text.lines()
        .map(|line| get_numbers(line, &number_map))
        .map(|line| get_edge_chars(line.as_str()))
        .map(|(first, last)| first * 10 + last).sum();
    
    println!("Result: {}", result)
}

fn get_numbers(input: &str, number_map: &HashMap<&str, &str>) -> String {
    let numbers_regex_string: &str = &format!("(?=(1{}))", 
    number_map.iter()
    .fold("".to_owned(), |acc, (k, v)| format!("{}|{}|{}", acc, k, v)));

    return Regex::new(numbers_regex_string)
        .expect(&format!("'{}' should be a valid regex", numbers_regex_string))
        .captures_iter(input)
        .map(|c| c.at(1))
        .filter(|o| o.is_some())
        .map(|m| m.unwrap())
        .fold(String::new(), |accumulator, match_string| {
            if match_string.len() == 1 {
                return accumulator + match_string
            } else {
                return accumulator + number_map.get(match_string).unwrap_or(&match_string);
            }
        });


}

fn get_edge_chars(input: &str) -> (u32, u32) {
    if input.chars().count() < 1 
        { return (0, 0); }
    
    let mut chars = input.chars();
    let first_char: char = chars.next().unwrap();
    let last_char: char = chars.last().unwrap_or(first_char.clone());

    //println!("Chars: '{}', '{}'", first_char, last_char);

    return (first_char.to_digit(10).unwrap_or(0), 
        last_char.to_digit(10).unwrap_or(0));
}